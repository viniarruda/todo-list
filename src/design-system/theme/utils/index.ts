import { TokenObject, TransformedTokenObject } from './types'

export const transformTokensToThemeObject = <T extends TokenObject>(
  tokens: T,
): TransformedTokenObject<T> => {
  return Object.entries(tokens).reduce((acc, [key, value]) => {
    acc[key as keyof T] = { value: value as T[keyof T] }
    return acc
  }, {} as TransformedTokenObject<T>)
}
