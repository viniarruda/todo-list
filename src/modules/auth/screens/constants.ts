import { SignInFormData } from "@/modules/auth/components/FormProvider/types";

export const defaultValues: SignInFormData = {
  email: '',
  password: '',
}