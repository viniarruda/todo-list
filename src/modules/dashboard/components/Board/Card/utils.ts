import { Column } from '@/services/entities/Board'
import { Todo } from '@/services/entities/Todo'
import { UpdateBoardCardsOrder, UpdateBoardColumns } from './types'

export const updateBoardCardsOrder = ({
  todos,
  sourceIndex,
  destinationIndex,
}: UpdateBoardCardsOrder) => {
  const updatedTodos = [...todos]

  const [movedTodo] = updatedTodos.splice(sourceIndex, 1)

  updatedTodos.splice(destinationIndex, 0, movedTodo)

  return updatedTodos
}

// export const updateBoardColumns = ({
//   columnTitle,
//   reorderedTodos,
//   columns,
// }: UpdateBoardColumns): Column[] => {
//   return (columns || []).map(column =>
//     column.title === columnTitle
//       ? { ...column, todos: reorderedTodos }
//       : column,
//   )
// }

export const updateBoardColumns = ({
  columns,
  draggedListIndex,
  targetListIndex,
  sourceIndex,
  destinationIndex,
}: UpdateBoardColumns): Column[] => {
  const updatedColumns = [...(columns || [])]

  // If moving within the same column, just reorder the todos
  if (draggedListIndex === targetListIndex) {
    const column = updatedColumns[draggedListIndex]
    column.todos = updateBoardCardsOrder({
      todos: column.todos,
      sourceIndex,
      destinationIndex,
    })
    return updatedColumns
  }

  // Moving across different columns
  const sourceColumn = updatedColumns[draggedListIndex]
  const targetColumn = updatedColumns[targetListIndex]

  const [movedTodo] = sourceColumn.todos.splice(sourceIndex, 1)

  // Insert the todo into the target column at the specified destination index
  targetColumn.todos.splice(destinationIndex, 0, movedTodo)

  return updatedColumns
}
