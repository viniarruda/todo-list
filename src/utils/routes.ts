export const routes = {
  HOME: '/',
  LOGIN: '/login/',
  REGISTER: '/register/',
  DASHBOARD: '/dashboard/',
  TODO_LISTS: '/dashboard/todo-lists/',
  NEW_TODO_LISTS: '/dashboard/todo-lists/new/',
}
