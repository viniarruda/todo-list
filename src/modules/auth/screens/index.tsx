'use client'

import { Flex } from '@design-system/components'

import { FormProvider } from '@/modules/auth/components/FormProvider'
import { FormRoot } from '@/modules/auth/components/FormRoot'

import { defaultValues } from './constants'

export const LoginScreen = () => {
  return (
    <Flex
      width="full"
      height="dvh"
      justify="center"
      align="center"
      bg="primary"
    >
      <FormProvider defaultValues={defaultValues}>
        <FormRoot />
      </FormProvider>
    </Flex>
  )
}
