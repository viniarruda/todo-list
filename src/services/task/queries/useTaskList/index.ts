import { UseQueryOptions, useQuery } from '@tanstack/react-query'

import { AxiosError } from 'axios'

import { microservices } from '@/services/constants'

import { createUseTaskListKey } from './key'
import { taskList } from './request'
import { TaskListData } from './types'

export const useTaskList = (
  options?: UseQueryOptions<TaskListData, AxiosError>,
) => {
  return useQuery<TaskListData, AxiosError>({
    queryKey: createUseTaskListKey(),
    queryFn: () => taskList(microservices.tasks),
    ...options,
  })
}
