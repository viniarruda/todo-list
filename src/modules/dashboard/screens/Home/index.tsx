'use client'

import { useBoardList } from '@/services/todo/queries/useTodoList'
import { Button, Flex, Title, Typography } from '@design-system/components'

export const Home = () => {
  const { data, isLoading } = useBoardList()

  if (isLoading) {
    return <div>Loading...</div>
  }

  if (!data?.length) {
    return (
      <div>
        <div>No Todo List registered!</div>
        <Button>Create new!</Button>
      </div>
    )
  }

  return (
    <Flex direction="column">
      <Title color="textPrimary">Todo lists</Title>
      <Flex wrap="wrap" bg="backgroundTertiary" borderRadius="sm" p="2">
        {data.map(board => (
          <div key={board.id}>
            <Typography fontSize="2xl" align="center">
              {board.title}
            </Typography>
            <Typography fontSize="md">{board.description}</Typography>
            {board.todos?.map(todo => (
              <div key={todo.id}>
                <Typography fontSize="2xl" align="center">
                  {todo.title}
                </Typography>
                <Typography fontSize="md">{todo.description}</Typography>
                <Flex bg="feedbackError">
                  <Typography color="textPrimary">{todo.status}</Typography>
                </Flex>
              </div>
            ))}
          </div>
        ))}
      </Flex>
    </Flex>
  )
}
