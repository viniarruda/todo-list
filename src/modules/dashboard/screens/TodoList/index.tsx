'use client'

import { Flex, Title, Typography } from '@/design-system/components'
import { useBoard } from '@/services/todo/queries/useTodo'
import { createUseBoardKey } from '@/services/todo/queries/useTodo/key'

import { List } from '@/modules/dashboard/components/Board/List'
import { Board } from '@/modules/dashboard/components/Board/styles'

import { useBoardStore } from '@/stores/useBoardStore'
import { useEffect } from 'react'
import { TodoListScreenProps } from './types'

export const TodoListScreen = ({ id }: TodoListScreenProps) => {
  const saveNewBoard = useBoardStore(state => state.saveNewBoard)

  const { data, isLoading } = useBoard({
    id,
  })

  useEffect(() => {
    if (data) {
      saveNewBoard(data)
    }
  }, [data])

  return (
    <Flex direction="column" width="full">
      <Title>Todo List</Title>
      {isLoading && <Flex>Loading...</Flex>}

      {data && (
        <Flex direction="column" width="full" height="full" color="inverse">
          <Flex direction="column" align="center">
            <Typography fontSize="5xl" color="textPrimary">
              {data.title}
            </Typography>
            <Typography fontSize="3xl" fontWeight="light" color="textPrimary">
              {data.description}
            </Typography>
          </Flex>

          {data.columns && (
            <Board>
              {data.columns.map((column, index) => (
                <List
                  key={column.title}
                  column={column}
                  listIndex={index}
                  id={id}
                />
              ))}
            </Board>
          )}
        </Flex>
      )}
    </Flex>
  )
}
