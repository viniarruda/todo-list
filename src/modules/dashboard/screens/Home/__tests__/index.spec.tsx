import { Home } from '@/modules/dashboard/screens/Home'
import { render, screen } from '@tests/utils/render'

import { routes } from '@/utils/routes'
import userEvent from '@testing-library/user-event'

import { useBoardList } from '@/services/task/queries/useTaskList'
import { responseMock } from './mock'

const pushMock = jest.fn()

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({
    push: pushMock,
  })),
}))

jest.mock('@/services/todo/queries/useTodoList', () => ({
  useBoardList: jest.fn(() => ({
    data: undefined,
    isLoading: false,
  })),
}))

const useBoardListMock = useBoardList as jest.Mock

describe('Dashboard home', () => {
  describe('when screen render', () => {
    it('should render correctly', () => {
      const { container } = render(<Home />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('when click to Go to Board', () => {
    it('should navigate to board', async () => {
      useBoardListMock.mockReturnValue({
        data: responseMock,
        isLoading: false,
      })

      render(<Home />)

      const button = screen.getByText('Go to Board')

      await userEvent.click(button)

      expect(pushMock).toHaveBeenCalledWith(routes.DASHBOARD_TODO_LISTS('123'))
    })
  })
})
