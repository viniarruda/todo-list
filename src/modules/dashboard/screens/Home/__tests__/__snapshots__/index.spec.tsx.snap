// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard home when screen render should render correctly 1`] = `
<div>
  <div>
    <div>
      No Todo List registered!
    </div>
    <button
      class="cursor_pointer bdr_sm d_flex jc_center ai_center [&:disabled]:cursor_not-allowed [&:disabled]:op_0.5 bg_backgroundTertiary c_textPrimary [&:hover]:bg_backgroundSecondary h_10 px_24 fs_md"
      type="submit"
    >
      Create new!
    </button>
  </div>
</div>
`;
